File: JavaScript/Array_Hashing/IsAnagram.js
/**
     * @param {string} s
     * @param {string} t
     * @return {boolean}
     * Given two strings s and t, return true if the two strings are anagrams of each other, 
     * otherwise return false. An anagram is a string that contains the exact same characters 
     * as another string, but the order of the characters can be different.
     */
function isAnagram(s, t) {
    if (s.length !== t.length) {
        return false
    }

    let stringInS = new Map()
    let stringInT = new Map()

    for (let i = 0; i < s.length; i++) {
        let charS = s[i];
        let charT = t[i];

        stringInS.set(charS, (stringInS.get(charS) || 0) + 1);
        stringInT.set(charT, (stringInT.get(charT) || 0) + 1);
    }

    for (let [key, value] of stringInS) {
        if (stringInT.get(key) !== value) {
            return false;
        }
    }

    return true;
}

console.log(isAnagram("listen", "silent")); // Output: true
console.log(isAnagram("hello", "world"));   // Output: false

File: C++/Array_Hashing/isAnagram.cpp
#include <vector>
#include <unordered_set>
#include <iostream>
using namespace std;

bool isAnagram(string s, string t) {
    unordered_map<char, int> stringInS;
    unordered_map<char, int> stringInT;

    // Count the frequency of each character in both strings
    for (char letter : s) {
        stringInS[letter]++;
    }
    for (char letter : t) {
        stringInT[letter]++;
    }
    // Compare the frequency maps
    return stringInS == stringInT;
}


int main() {
    string s = "anagram";
    string t = "nagaram";
    
    if (isAnagram(s, t)) {
        cout << "The strings are anagrams." << endl;
    } else {
        cout << "The strings are not anagrams." << endl;
    }
    return 0;
}